Mathlib/Tactic/TFAE.lean

```diff
@@ -159,62 +223,89 @@ def proveTFAE (is : List ℕ) (l : Q(List Prop)) : MetaM Q(TFAE $l) := do
-    let i' ← elabIndex i l₀
-    let j' ← elabIndex j l₀
+/-- Accesses the propositions at indices `i` and `j` of `tfaeList`, and constructs the expression
+`Pi <arr> Pj`, which will be the type of our `tfae_have` hypothesis -/
+def elabTFAEType (tfaeList : List Q(Prop)) : TSyntax ``tfaeType → TermElabM Expr
+  | stx@`(tfaeType|$i:num $arr:impArrow $j:num) => do
+    let l := tfaeList.length
+    let i' ← elabIndex i l
+    let j' ← elabIndex j l
     let Pi := tfaeList.get! (i'-1)
     let Pj := tfaeList.get! (j'-1)
-    let type ← mkImplType Pi arr Pj
-    let (goal1, goal2) ← tfaeHaveCore goal h i j arr type
-    replaceMainGoal [goal1, goal2]
+    Term.addTermInfo' i Pi q(Prop)
+    Term.addTermInfo' j Pj q(Prop)
+    match arr with
+    | `(impArrow| ← ) => Term.addTermInfo stx q($Pj → $Pi) q(Prop)
+    | `(impArrow| → ) => Term.addTermInfo stx q($Pi → $Pj) q(Prop)
+    | `(impArrow| ↔ ) => Term.addTermInfo stx q($Pi ↔ $Pj) q(Prop)
+    | _ => throwUnsupportedSyntax
+  | _ => throwUnsupportedSyntax
+
+/- Convert `tfae_have i <arr> j ...` to `tfae_have tfae_i_arr_j : i <arr> j ...`. See
+`expandHave`, which is responsible for inserting `this` in `have : A := ...`. Note that we
+require some extra help for `tfaeHave'` (Mathlib `have`). -/
+macro_rules
+| `(tfaeHave|tfae_have $hy:hygieneInfo $t:tfaeType := $val) => do
+  let id := HygieneInfo.mkIdent hy (← mkTFAEId t) (canonical := true)
+  `(tfaeHave|tfae_have $id : $t := $val)
+| `(tfaeHave|tfae_have $hy:hygieneInfo $t:tfaeType $alts:matchAlts) => do
+  let id := HygieneInfo.mkIdent hy (← mkTFAEId t) (canonical := true)
+  `(tfaeHave|tfae_have $id : $t $alts)
+
+-- Mathlib `have`
+| `(tfaeHave'|tfae_have $hy:hygieneInfo $t:tfaeType) => do
+  let id := HygieneInfo.mkIdent hy (← mkTFAEId t) (canonical := true)
+  `(tfaeHave'|tfae_have $id : $t)
+
+open Term
+
+elab_rules : tactic
+| `(tfaeHave|tfae_have $d:tfaeHaveDecl) => withMainContext do
+  let goal ← getMainGoal
+  let (_, tfaeList) ← getTFAEList (← goal.getType)
+  withRef d do
+    match d with
+    | `(tfaeHaveDecl| $b : $t:tfaeType := $pf:term) =>
+      let type ← elabTFAEType tfaeList t
+      evalTactic <|← `(tactic|have $b : $(← exprToSyntax type) := $pf)
+    | `(tfaeHaveDecl| $b : $t:tfaeType $alts:matchAlts) =>
+      let type ← elabTFAEType tfaeList t
+      evalTactic <|← `(tactic|have $b : $(← exprToSyntax type) $alts:matchAlts)
+    | `(tfaeHaveDecl| $pat:term : $t:tfaeType := $pf:term) =>
+      let type ← elabTFAEType tfaeList t
+      evalTactic <|← `(tactic|have $pat:term : $(← exprToSyntax type) := $pf)
+    | _ => throwUnsupportedSyntax
+
+-- Mathlib `have`
+| `(tfaeHave'|tfae_have $d:tfaeHaveIdLhs) => withMainContext do
+  let goal ← getMainGoal
+  let (_, tfaeList) ← getTFAEList (← goal.getType)
+  match d with
+  | `(tfaeHaveIdLhs| $b:ident : $t:tfaeType) =>
+    let type ← elabTFAEType tfaeList t
+    evalTactic <|← `(tactic|have $b:ident : $(← exprToSyntax type))
+  | _ => throwUnsupportedSyntax
```

Comment:
I think the current consensus is that the mathlib-`have` should not be used inside mathlib anymore, and therefore ideally also not imported (see [docs](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Tactic/Have.html)):

> As a style choice, this [mathlib's `have`] should not be used in mathlib; but is provided for downstream users who preferred the old style.

Maybe it's easiest to just drop the import together with this snippet?

Or is there a reasonable workaround which allows not to import the modified `have` here?

Reply:
Ah, that's right. :) I eliminated the import and moved the entirety of the old-style version to its own section below.

In parallel to Mathlib `have`, I think we ought to move it to its own non-imported file in the PR that migrates Mathlib to the new syntax (#11000); what do you think?